### PR TITLE
nGraph shell for the operation IDFT

### PR DESCRIPTION
**Pull request by vgavrilo**
### Details:
 - *nGraph shell for the operation IDFT (Inverse Discrete Fourier Transform)*

### Tickets:
 - *49370*

### CODE REVIEW

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (https://github.com/openvinotoolkit/openvino/pull/4660)

**Review by ilyachur — COMMENTED** [review]
The same comments as for PR https://github.com/openvinotoolkit/openvino/pull/4625

**Comment by vgavrilo** [comment]
> The same comments as for PR #4625

All these comments were fixed.

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

File: ngraph/core/src/op/idft.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    PartialShape input_shape = PartialShape(get_input_partial_shape(0));
+    PartialShape axes_shape = PartialShape(get_input_partial_shape(1));
+    PartialShape output_shape = input_shape;
+
+    if (input_shape.rank().is_dynamic() || axes_shape.rank().is_dynamic() ||
+        !is_type<op::Constant>(input_value(1).get_node()))
+    {
+        set_output_type(0, get_input_element_type(0), output_shape);
+        return;
+    }
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/idft.cpp:100]: If the `signal_size` input is specified (and not all elements are equal to `-1`) then the `output_shape` is not equal to the `input_shape`!  So this check is invalid.

If the `signal_size` input is not specified then we can should `input_shape` as `output_shape`. In the other case we should use values from `axes` input and `signal_size` and calculate output shape.

If the `axes` values are not specified then the output shape tensor should have all elements to be Dynamic, except the last one which should be equal to 2.

File: ngraph/core/src/op/idft.cpp [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        }
+    }
+
+    NODE_VALIDATION_CHECK(this,
+                          input_rank >= axes.size() + 1,
+                          "The input rank must be greater than number of IDFT axes. Got "
+                          "input rank: ",
+                          input_rank,
+                          ", number of axes: ",
+                          axes.size());
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/idft.cpp:132]: Please, separate logic of checking inputs shapes and values from the output shape calculation.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/idft.cpp (hunk 1 above)
Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp:100]: But we cannot calculate output shape using `signal_size`, if `axes` or `signal_size` are not constants. For eaxmple, `signal_size` is a constant with not all elements are equal to `-1`, but `axes` is not constant.

File: ngraph/core/src/op/idft.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    NODE_VALIDATION_CHECK(this,
+                          input_et == element::f32 || input_et == element::f16 ||
+                              input_et == element::bf16,
+                          "Input element type must be f32, f16, or bf16");
+
+    element::Type axes_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          axes_et == element::i64 || axes_et == element::i32 ||
+                              axes_et == element::u32 || axes_et == element::u64,
+                          "Axes element type must be i32, i64, u32 or u64");
```
[the remark this answers is not in the document]
Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: Fixed.

File: ngraph/core/src/op/idft.cpp (hunk 1 above)
Comment (lazarevevgeny) [reply ngraph/core/src/op/idft.cpp]: Yes, we cannot calculate them. But we cannot set the output shape to be equal to `input_shape`. The output shape should be {Dynamic, Dynamic,...,Dynamic, 2}.

File: ngraph/core/src/op/idft.cpp (hunk 2 above)
Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: Done.

File: ngraph/core/src/op/idft.cpp (hunk 1 above)
Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: And then, if `axes` is constant and `signal_size` is not constant and has static rank, we should set elements of `output_shape` specified in `axes` as `dynamic()`.

Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: Done.

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

File: ngraph/core/src/op/idft.cpp [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        NODE_VALIDATION_CHECK(
+            this, axes.size() == axes_set.size(), "IDFT axes must be unique. Got: ", axes_vector);
+
+        NODE_VALIDATION_CHECK(this,
+                              std::find(axes.begin(), axes.end(), input_rank - 1) == axes.end(),
+                              "IDFT axes cannot contain the last axis. Got axes: ",
+                              axes_vector);
+    }
+
+    if (num_of_inputs == 3 && axes_shape.is_static() && get_input_partial_shape(2).is_static())
```
Comment (ilyachur) [inline ngraph/core/src/op/idft.cpp:173]: Can we try to merge `if`s for `num_of_inputs == 3` condition?

Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: Done.

**Comment by vgavrilo** [comment]
@mvafin, @evolosen, @yekruglov, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Review by evolosen — APPROVED** [review]

File: ngraph/core/src/op/idft.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+                                  "IDFT Signal size input must be 1D tensor. Got signal size "
+                                  "input rank: ",
+                                  signal_size_shape.rank().get_length());
+        }
+
+        if (axes_shape.is_static() && signal_size_shape.is_static())
+        {
+            NODE_VALIDATION_CHECK(this,
+                                  axes_shape.to_shape()[0] == signal_size_shape.to_shape()[0],
+                                  "Sizes of inputs 'axes' and 'sinal_size' must be equal. Got "
```
Comment (yekruglov) [inline ngraph/core/src/op/idft.cpp:176]: There is a typo in `signal_size`

**Review by yekruglov — APPROVED** [review]

File: ngraph/core/src/op/idft.cpp (hunk 5 above)
Comment (vgavrilo) [reply ngraph/core/src/op/idft.cpp]: Fixed.

**Comment by vgavrilo** [comment]
@mvafin, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
@mvafin, @lazarevevgeny, @ilyachur, please, see this PR (#4660)

**Review by ilyachur — APPROVED** [review]

**Comment by vgavrilo** [comment]
@mvafin, @lazarevevgeny, please, see this PR (#4660)

**Review by lazarevevgeny — APPROVED** [review]

**Comment by lazarevevgeny** [comment]
@vgavrilo , please, resolve merge conflict

**Comment by vgavrilo** [comment]
@mvafin, please, see this PR (#4660)

**Comment by vgavrilo** [comment]
> @vgavrilo , please, resolve merge conflict

Done.

**Pull request merged**